src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs

```diff
@@ -401,21 +401,253 @@ internal List<PSModuleInfo> GetModules(ModuleSpecification[] fullyQualifiedName,
+        /// <param name="moduleInfo">The module info object to check.</param>
+        /// <param name="name">The name of the expected module.</param>
+        /// <param name="guid">The guid of the expected module.</param>
+        /// <param name="requiredVersion">The required version of the expected module.</param>
+        /// <param name="minimumVersion">The minimum required version of the expected module.</param>
+        /// <param name="maximumVersion">The maximum required version of the expected module.</param>
+        /// <returns>True if the module info object matches all given constraints, false otherwise.</returns>
+        internal static bool IsModuleMatchingConstraints(
+            out ModuleMatchFailure matchFailureReason,
+            PSModuleInfo moduleInfo,
+            string name = null,
+            Guid? guid = null,
+            Version requiredVersion = null,
+            Version minimumVersion = null,
+            Version maximumVersion = null)
+        {
+            // Define that a null module does not meet any constraints
+            if (moduleInfo == null)
+            {
+                matchFailureReason = ModuleMatchFailure.NullModule;
+                return false;
+            }
+
+            return AreModuleFieldsMatchingConstraints(
+                out matchFailureReason,
+                moduleInfo.Name,
+                moduleInfo.Guid,
+                moduleInfo.Version,
+                name,
+                guid,
+                requiredVersion,
+                minimumVersion,
+                maximumVersion
+            );
+        }
+
+        /// <summary>
+        /// Check that given module fields meet any given constraints.
+        /// </summary>
+        /// <param name="moduleName">The name of the module to check.</param>
+        /// <param name="moduleGuid">The GUID of the module to check.</param>
+        /// <param name="moduleVersion">The version of the module to check.</param>
+        /// <param name="requiredName">The name the module must have, if any.</param>
+        /// <param name="requiredGuid">The GUID the module must have, if any.</param>
+        /// <param name="requiredVersion">The exact version the module must have, if any.</param>
+        /// <param name="minimumRequiredVersion">The minimum version the module may have, if any.</param>
+        /// <param name="maximumRequiredVersion">The maximum version the module may have, if any.</param>
+        /// <returns>True if the module parameters match all given constraints, false otherwise.</returns>
+        internal static bool AreModuleFieldsMatchingConstraints(
+            string moduleName = null,
+            Guid? moduleGuid = null,
+            Version moduleVersion = null,
+            string requiredName = null,
+            Guid? requiredGuid = null,
+            Version requiredVersion = null,
+            Version minimumRequiredVersion = null,
+            Version maximumRequiredVersion = null)
```

Comment:
Does it really make sense to be so permissive?

Reply:
On the one hand, I don't think we need all the overloads I've included here currently, or at least not taking so many parameters.

On the other hand, the parameter set matches the problem domain:
- We have a `ModuleInfo` or its fields. In some cases we have no name, in others we have no guid, and we are never guaranteed that the version is not `null` by the caller
- We are given a set of module constraints (usually a `ModuleSpecification`) which we model as a hashtable with a number of fields that may or may not be there
- The method's contract is basically to take whatever parts of the module it knows about and see if they fit within whatever constraints it was given
- If there are no constraints, then anything matches

Does this make the behaviour permissive? Possibly. But I think it's well-reasoned and consistent, meaning we can depend on it in any future scenarios, rather than setting its behaviour to suit the current scenario and having to break it later.